refactor: error messages reformatted and moved into own files (#1731)

fn-faisal opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
Message error for required fields #1731

Related issue: [Issue link](https://github.com/parse-community/parse-dashboard/issues/1731)

### Approach
This is a PR regarding the below issue. The formatting has been added to the server. I'll update the existing PR for the dashboard to exclude formatting of response string from dashboard. 
[https://github.com/parse-community/parse-dashboard/issues/1731#issuecomment-865070353](https://github.com/parse-community/parse-dashboard/issues/1731#issuecomment-865070353)

faisal154 committed:
Error message formatting on server

mtrezza commented on src/middlewares.js:
```diff
@@ -360,8 +361,15 @@ export function handleParseErrors(err, req, res, next) {
         httpStatus = 400;
     }
+    // capitalize the message if the error code isn't missing field field
+    if (err.code !== 142) {
```
I'm not sure we should introduce a change in capitalization for a single error message. Not only may this be a breaking change for some deployments, but it is also missing a broader concept of how error messages should be styled across the server.

mtrezza commented:
Not sure about this PR, see https://github.com/parse-community/parse-dashboard/issues/1731#issuecomment-868657186.

faisal154 committed:
removed formatting issue

fn-faisal commented:
I have updated this PR to refactor error messages from server into their own files. This way, the error messages are easier to update. I have also updated the error message for the required fields.

fn-faisal committed:
Merge branch 'master' into feat/1731

fn-faisal commented:
@mtrezza For this one, I refactored the messages in the a separate file and refactored the messages. I wasn't too sure about the changes required for this issue but from what I understood, I refactored the message strings in their own file and referenced them in the error responses.
I didn't see a support for locale strings and wasn't sure if multiple locales were supported so I made a new file for locale string.

mtrezza commented:
That's certainly a nice consolidation!

- There are still some error messages that should begin uppercase, I think, for example `'bad or missing username'`
- Could you rename the message.js path to be more specific, for example `errors.js`?
- For the tests to pass, they should also use the error references.

faisal154 committed:
refactor -- 1

faisal154 committed:
Merge branch 'feat/1731' of https://github.com/fn-faisal/parse-server into feat/1731

faisal154 committed:
refactor -- 2

faisal154 committed:
refactor -- 3

faisal154 committed:
refactor -- 4

mtrezza committed:
Merge branch 'master' into feat/1731

mtrezza commented:
@fn-faisal Is this ready for review?

fn-faisal committed:
Merge branch 'master' into feat/1731

fn-faisal commented:
@mtrezza Yes, it's ready for review.

mtrezza committed:
Merge branch 'master' into feat/1731

mtrezza commented:
Could you add a changelog entry?

faisal154 committed:
CHANGELOG entry + removed . at the end of errors

faisal154 committed:
updated specs

faisal154 committed:
updated error message 2

fn-faisal committed:
Merge branch 'master' into feat/1731

mtrezza commented:
Can you give an example when it would be odd to see a dot at the end of an error message?

fn-faisal commented:
@mtrezza I agree. I'll update the messages to have a dot at the end.

faisal154 committed:
Added dot at the end of messages

faisal154 committed:
dot at the end of messages -- 2

mtrezza commented:
Alright, sounds good

mtrezza committed:
Merge branch 'master' into feat/1731

mtrezza committed:
Merge branch 'master' into feat/1731

mtrezza commented:
Thanks, could you take a look at the failing tests?

faisal154 committed:
url comparison spec file fix

mtrezza commented:
It seems to pass! Please request a review whenever this is ready.

faisal154 committed:
fixing spec -- 2

faisal154 committed:
fixing spec -- 3

mtrezza reviewed (commented):
Wow, that's some work! Well done, just found a few things...

faisal154 committed:
fixed error messages

faisal154 committed:
Merge branch 'master' of https://github.com/parse-community/parse-server into feat/1731

mtrezza committed:
Merge branch 'master' into feat/1731

mtrezza commented:
@faisal154 is this ready for review?

mtrezza committed:
Merge branch 'master' into feat/1731

mtrezza committed:
Merge branch 'master' into feat/1731

mtrezza commented:
@fn-faisal It seems that some tests are still failing, could you take a look? And maybe comment on the open review questions, so we know which ones are done. I would rather merge his soon, otherwise there may be other changes in the meantime that make it more difficult to merge.

fn-faisal commented:
@mtrezza sorry, there were some urgent issues at work. I'll start looking into resolving the test cases.

faisal154 committed:
fixed error

faisal154 committed:
Merge branch 'feat/1731' of https://github.com/fn-faisal/parse-server into feat/1731

faisal154 committed:
test fixes -- 2

mtrezza commented:
Great, no worries, whenever you are ready!

faisal154 committed:
error fix

faisal154 committed:
error message revert

faisal154 committed:
error message revert -- 2

faisal154 committed:
fix error messages

fn-faisal commented:
@mtrezza I'm seeing `notContainedIn false queries` test pass for mongoDB but fail for Postgres. It fails saying the error message should contain a `.` at the end but the message, I didn't reformat in code as I couldn't find any reference for it, except in tests. This may be in the SDK. And the message is different for MongoDb tests and Postgres tests that's why the tests fail.

faisal154 committed:
updated testcase for notContainedIn false queries

faisal154 committed:
resolve merge conflicts

faisal154 committed:
fix typo

faisal154 committed:
workarounds for postgres tests

faisal154 committed:
Merge branch 'alpha' of https://github.com/parse-community/parse-server into feat/1731

faisal154 committed:
workaround 1

faisal154 committed:
fixes for pgsql tests

faisal154 committed:
fixes for pgsql tests

faisal154 committed:
Merge branch 'alpha' of https://github.com/parse-community/parse-server into feat/1731

fn-faisal committed:
Merge branch 'alpha' into feat/1731

fn-faisal commented:
@mtrezza I have added checks for the tests that were failing for POSTGRES. Please let me know if the PR requires any updates.

mtrezza commented on spec/ParseGraphQLServer.spec.js:
```diff
@@ -2600,11 +2600,13 @@ describe('ParseGraphQLServer', () => {
-              ).toBeLessThan(
-                findSecondaryObjectsResult.data.secondaryObjects.edges[1].node.objectId
-              );
+              if (process.env.PARSE_SERVER_TEST_DB != 'postgres') {
```
Unfortunately, this is not how tests are exluded for certain DB adapters, can you please remove all the:

```
if (process.env.PARSE_SERVER_TEST_DB != 'postgres') {
```

There is a `it_only_mongo`, etc. syntax that is used. However, this should not be part of this PR, because this PR focuses on refactoring error messages. Changing tests is a delicate thing that needs to be handled in a separate issue and PR. I suggest we focus this PR on the error messages and open a new issue for the test changes to discuss.

Could you please request a review again once the lines have been removed and if all error messages start capitalized and end with a period, this should be good to merge.

faisal154 committed:
removed if checks for postgres

faisal154 committed:
Merge branch 'feat/1731' of https://github.com/fn-faisal/parse-server into feat/1731

faisal154 committed:
fix: error message type fix

fn-faisal commented:
@mtrezza I have reverted the changes and created an issue for the failing tests. [https://github.com/parse-community/parse-server/issues/7732](https://github.com/parse-community/parse-server/issues/7732)

mtrezza commented on spec/ParseQuery.spec.js:
```diff
@@ -980,7 +980,9 @@ describe('Parse.Query testing', () => {
         equal(response.data.code, Parse.Error.INVALID_JSON);
-        equal(response.data.error, 'bad $containedBy: should be an array');
+        expect(
+          ['bad $containedBy: should be an array', undefined].indexOf(response.data.error) > -1
```
Why can this not expect the full error message? If that's not possible, please use `expect(...).toContain(...)` instead of `indexOf`. Also in the tests below.

fn-faisal replied:
I have reverted this. I missed this when I was reverting the workaround for failing tests.

mtrezza commented on spec/ParseInstallation.spec.js:
```diff
@@ -1232,7 +1232,7 @@ describe('Installations', () => {
           err => {
             expect(err.code).toBe(136);
-            expect(err.message).toBe('installationId may not be changed in this operation');
+            expect(err.message).toBe('The installationId can not be changed in this operation.');
```
Please change to:

> The field installationId ...

To keep the same syntax that you used in the other messages.

fn-faisal replied:
I have updated it.

mtrezza commented on src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:
```diff
@@ -1387,7 +1387,7 @@ export class PostgresStorageAdapter implements StorageAdapter {
           const err = new Parse.Error(
             Parse.Error.DUPLICATE_VALUE,
-            'A duplicate value for a field with unique values was provided'
+            'A duplicate value for a field with unique values was provided.'
```
Shouldn't this be `ErrorMessage.DUPLICATE_VALUE())`?

mtrezza commented on src/Routers/UsersRouter.js:
```diff
@@ -76,7 +76,7 @@ export class UsersRouter extends ClassesRouter {
       }
       if (!password) {
-        throw new Parse.Error(Parse.Error.PASSWORD_MISSING, 'password is required.');
+        throw new Parse.Error(Parse.Error.PASSWORD_MISSING, 'Password is required.');
```
Should this also use ErrorMessage instead of a string?

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+export const ErrorMessage = {
+  READ_ONLY_MASTER_KEY: () => 'Cannot perform a write operation when using readOnlyMasterKey.',
```
masterKeyReadOnly

fn-faisal replied:
I have updated this.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+export const ErrorMessage = {
+  READ_ONLY_MASTER_KEY: () => 'Cannot perform a write operation when using readOnlyMasterKey.',
+  MISSING_OBJECT_ID: () => 'The field objectId must not be empty, null or undefined.',
```
objectIdMissing

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+export const ErrorMessage = {
+  READ_ONLY_MASTER_KEY: () => 'Cannot perform a write operation when using readOnlyMasterKey.',
+  MISSING_OBJECT_ID: () => 'The field objectId must not be empty, null or undefined.',
+  INVALID_KEY_NAME: value => value + ' is an invalid field name.',
```
objectKeyNameInvalid

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  READ_ONLY_MASTER_KEY: () => 'Cannot perform a write operation when using readOnlyMasterKey.',
+  MISSING_OBJECT_ID: () => 'The field objectId must not be empty, null or undefined.',
+  INVALID_KEY_NAME: value => value + ' is an invalid field name.',
+  USER_NO_ACCESS_CLASS: className =>
```
userPermissionClassAccessDenied

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  INVALID_KEY_NAME: value => value + ' is an invalid field name.',
+  USER_NO_ACCESS_CLASS: className =>
+    'This user is not allowed to access non-existent class: ' + className + '.',
+  OBJECT_NOT_FOUND: () => 'Object not found.',
```
objectNotFound

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  USER_NO_ACCESS_CLASS: className =>
+    'This user is not allowed to access non-existent class: ' + className + '.',
+  OBJECT_NOT_FOUND: () => 'Object not found.',
+  FIELD_IS_REQUIRED: fieldName =>
```
objectFieldValueRequired

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  OBJECT_NOT_FOUND: () => 'Object not found.',
+  FIELD_IS_REQUIRED: fieldName =>
+    'A value for the field ' + fieldName + ' is required to save/update the object.',
+  USERNAME_MISSING: () => 'Bad or missing username.',
```
userUsernameMissing

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  FIELD_IS_REQUIRED: fieldName =>
+    'A value for the field ' + fieldName + ' is required to save/update the object.',
+  USERNAME_MISSING: () => 'Bad or missing username.',
+  PASSWORD_MISSING: () => 'Password is required.',
```
userPasswordMissing

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+    'A value for the field ' + fieldName + ' is required to save/update the object.',
+  USERNAME_MISSING: () => 'Bad or missing username.',
+  PASSWORD_MISSING: () => 'Password is required.',
+  UNSUPPORTED_SERVICE: () => 'This authentication method is unsupported.',
```
authUnsupportedService

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  USERNAME_MISSING: () => 'Bad or missing username.',
+  PASSWORD_MISSING: () => 'Password is required.',
+  UNSUPPORTED_SERVICE: () => 'This authentication method is unsupported.',
+  ACCOUNT_LINKED: () => 'This auth is already used.',
```
authAccountLinked

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  PASSWORD_MISSING: () => 'Password is required.',
+  UNSUPPORTED_SERVICE: () => 'This authentication method is unsupported.',
+  ACCOUNT_LINKED: () => 'This auth is already used.',
+  CLIENT_EMAIL_VERIFICATION: () => "Clients aren't allowed to manually update email verification.",
```
userEmailVerificationClientDenied

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  UNSUPPORTED_SERVICE: () => 'This authentication method is unsupported.',
+  ACCOUNT_LINKED: () => 'This auth is already used.',
+  CLIENT_EMAIL_VERIFICATION: () => "Clients aren't allowed to manually update email verification.",
+  USERNAME_TAKEN: () => 'Account already exists for this username.',
```
userUsernameTaken

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  ACCOUNT_LINKED: () => 'This auth is already used.',
+  CLIENT_EMAIL_VERIFICATION: () => "Clients aren't allowed to manually update email verification.",
+  USERNAME_TAKEN: () => 'Account already exists for this username.',
+  INVALID_EMAIL_ADDRESS_FORMAT: () => 'Email address format is invalid.',
```
userEmailFormatInvalid

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  CLIENT_EMAIL_VERIFICATION: () => "Clients aren't allowed to manually update email verification.",
+  USERNAME_TAKEN: () => 'Account already exists for this username.',
+  INVALID_EMAIL_ADDRESS_FORMAT: () => 'Email address format is invalid.',
+  EMAIL_TAKEN: () => 'Account already exists for this email address.',
```
userEmailTaken

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  USERNAME_TAKEN: () => 'Account already exists for this username.',
+  INVALID_EMAIL_ADDRESS_FORMAT: () => 'Email address format is invalid.',
+  EMAIL_TAKEN: () => 'Account already exists for this email address.',
+  PASSWORD_POLICY: () => 'Password does not meet the Password Policy requirements.',
```
userPasswordPolicyFailed

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  INVALID_EMAIL_ADDRESS_FORMAT: () => 'Email address format is invalid.',
+  EMAIL_TAKEN: () => 'Account already exists for this email address.',
+  PASSWORD_POLICY: () => 'Password does not meet the Password Policy requirements.',
+  USERNAME_IN_PASSWORD: () => 'Password cannot contain your username.',
```
userUsernameInPassword

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  EMAIL_TAKEN: () => 'Account already exists for this email address.',
+  PASSWORD_POLICY: () => 'Password does not meet the Password Policy requirements.',
+  USERNAME_IN_PASSWORD: () => 'Password cannot contain your username.',
+  PASSWORD_MATCHES_EXISTING_PASSWORD: maxPasswordHistory =>
```
userPasswordMatchesPreviousPassword

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  USERNAME_IN_PASSWORD: () => 'Password cannot contain your username.',
+  PASSWORD_MATCHES_EXISTING_PASSWORD: maxPasswordHistory =>
+    'New password should not be the same as last ' + maxPasswordHistory + ' passwords.',
+  INVALID_SESSION_TOKEN: () => 'Session token required.',
```
sessionTokenInvalid

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  PASSWORD_MATCHES_EXISTING_PASSWORD: maxPasswordHistory =>
+    'New password should not be the same as last ' + maxPasswordHistory + ' passwords.',
+  INVALID_SESSION_TOKEN: () => 'Session token required.',
+  ACL_SESSION: () => 'Cannot set ' + 'ACL on a Session.',
```
sessionAclDenied

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+    'New password should not be the same as last ' + maxPasswordHistory + ' passwords.',
+  INVALID_SESSION_TOKEN: () => 'Session token required.',
+  ACL_SESSION: () => 'Cannot set ' + 'ACL on a Session.',
+  SESSION_ERROR: () => 'Error creating session.',
```
sessionCreationError

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  INVALID_SESSION_TOKEN: () => 'Session token required.',
+  ACL_SESSION: () => 'Cannot set ' + 'ACL on a Session.',
+  SESSION_ERROR: () => 'Error creating session.',
+  MIN_ONE_ID_FIELD: () =>
```
installationIdentifierRequired

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  SESSION_ERROR: () => 'Error creating session.',
+  MIN_ONE_ID_FIELD: () =>
+    'At least one ID field (deviceToken, installationId) must be specified in this operation.',
+  UNKNOWN_OBJECT_UPDATE: () => 'Object not found for update.',
```
I don't think we need a distinct error message for this; we can remove this message and use the generic `objectNotFound` instead.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  MIN_ONE_ID_FIELD: () =>
+    'At least one ID field (deviceToken, installationId) must be specified in this operation.',
+  UNKNOWN_OBJECT_UPDATE: () => 'Object not found for update.',
+  NO_OP: field => 'The ' + field + ' can not be changed in this operation.',
```
objectUpdateFieldFailed

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+    'At least one ID field (deviceToken, installationId) must be specified in this operation.',
+  UNKNOWN_OBJECT_UPDATE: () => 'Object not found for update.',
+  NO_OP: field => 'The ' + field + ' can not be changed in this operation.',
+  REQ_OP: field => 'The ' + field + ' must be specified in this operation.',
```
objectUpdateFieldRequired

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  UNKNOWN_OBJECT_UPDATE: () => 'Object not found for update.',
+  NO_OP: field => 'The ' + field + ' can not be changed in this operation.',
+  REQ_OP: field => 'The ' + field + ' must be specified in this operation.',
+  INSTALLATION_ID_WITH_DEVICE_TOKEN: () =>
```
installationIdRequiredForMultipleDeviceToken

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  REQ_OP: field => 'The ' + field + ' must be specified in this operation.',
+  INSTALLATION_ID_WITH_DEVICE_TOKEN: () =>
+    'Must specify installationId when deviceToken matches multiple Installation objects.',
+  USER_NO_MODIFY: field => 'Cannot modify user' + field + '.',
```
userUpdateFailed

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+    'Must specify installationId when deviceToken matches multiple Installation objects.',
+  USER_NO_MODIFY: field => 'Cannot modify user' + field + '.',
+  INVALID_ACL: () => 'Invalid ACL.',
+  DUPLICATE_VALUE: () => 'A duplicate value for a field with unique values was provided.',
```
objectValueDuplicate

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  USER_NO_MODIFY: field => 'Cannot modify user' + field + '.',
+  INVALID_ACL: () => 'Invalid ACL.',
+  DUPLICATE_VALUE: () => 'A duplicate value for a field with unique values was provided.',
+  DUPLICATE_CLASS: className => 'Class ' + className + 'already exists.',
```
classNameDuplicate

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  INVALID_ACL: () => 'Invalid ACL.',
+  DUPLICATE_VALUE: () => 'A duplicate value for a field with unique values was provided.',
+  DUPLICATE_CLASS: className => 'Class ' + className + 'already exists.',
+  DATABASE_ADAPTER_ERROR: () => 'Database adapter error.',
```
databaseAdapterError

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  DUPLICATE_VALUE: () => 'A duplicate value for a field with unique values was provided.',
+  DUPLICATE_CLASS: className => 'Class ' + className + 'already exists.',
+  DATABASE_ADAPTER_ERROR: () => 'Database adapter error.',
+  GEO_JSON_INVALID_VERTICES: () => 'GeoJSON: Loop must have at least 3 different vertices.',
```
geoJsonVerticesInvalid

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  DUPLICATE_CLASS: className => 'Class ' + className + 'already exists.',
+  DATABASE_ADAPTER_ERROR: () => 'Database adapter error.',
+  GEO_JSON_INVALID_VERTICES: () => 'GeoJSON: Loop must have at least 3 different vertices.',
+  INVALID_POLYGON_VALUES: () => 'Polygon must have atleast 3 values.',
```
geoPolygonMissingValues

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  DATABASE_ADAPTER_ERROR: () => 'Database adapter error.',
+  GEO_JSON_INVALID_VERTICES: () => 'GeoJSON: Loop must have at least 3 different vertices.',
+  INVALID_POLYGON_VALUES: () => 'Polygon must have atleast 3 values.',
+  INVALID_NESTED_KEY: () => "Nested keys should not contain the '$' or '.' characters",
```
objectNestedKeyNameInvalid

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  GEO_JSON_INVALID_VERTICES: () => 'GeoJSON: Loop must have at least 3 different vertices.',
+  INVALID_POLYGON_VALUES: () => 'Polygon must have atleast 3 values.',
+  INVALID_NESTED_KEY: () => "Nested keys should not contain the '$' or '.' characters",
+  BAD_VALUE: type => 'bad ' + type + 'value.',
```
objectFieldValueInvalid

fn-faisal replied:
I've made a general invalid field for this.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  INVALID_POLYGON_VALUES: () => 'Polygon must have atleast 3 values.',
+  INVALID_NESTED_KEY: () => "Nested keys should not contain the '$' or '.' characters",
+  BAD_VALUE: type => 'bad ' + type + 'value.',
+  ALL_VALUES_MUST_BE_REGEX: value =>
```
queryAllValueInvalid

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  BAD_VALUE: type => 'bad ' + type + 'value.',
+  ALL_VALUES_MUST_BE_REGEX: value =>
+    'All $all values must be of regex type or none: ' + value + '.',
+  VALUE_SHOULD_BE_TYPE: (value, type) => 'bad ' + value + 'should be an ' + type + '.',
```
`queryValueTypeInvalid: (key, type)` instead of `value, type`.

And the error message should probably be:
```
`Query key ${key} must be of type ${type}`
```

fn-faisal replied:
I have updated it.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  ALL_VALUES_MUST_BE_REGEX: value =>
+    'All $all values must be of regex type or none: ' + value + '.',
+  VALUE_SHOULD_BE_TYPE: (value, type) => 'bad ' + value + 'should be an ' + type + '.',
+  VALUE_NOT_SUPPORTED: value =>
```
Do we really need a distinct `value not supported` error message, or can we use a more generic error message, like the existing `queryValueTypeInvalid` or maybe a `queryValueInvalid`?

fn-faisal replied:
I have updated it.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,48 @@
+  VALUE_SHOULD_BE_TYPE: (value, type) => 'bad ' + value + 'should be an ' + type + '.',
+  VALUE_NOT_SUPPORTED: value =>
+    'bad ' + value + 'not supported, please use $regex or create a separate lower case column.',
+  BAD_VALUE_POSTGRES_EXTENSION_REQUIRED: value =>
```
```
databasePostgresExtensionRequired: (key, extension) => `Query key ${key} requires Postgres extension ${extension}.`,
```

fn-faisal replied:
I have updated this.

mtrezza reviewed (changes requested):
Here some general feedback regarding `src/Error/message.js`:

- all keys should be rewritten camelCased because they are functions, not constants, for example:
  ```
  readOnlyMasterKey() => 'Cannot perform a write operation when using readOnlyMasterKey.',
  ```
- to bring more order into the list and make it easier to find existing messages to prevent duplicates, let's define a syntax like `<subject><property><conflict>´. For example instead of `invalidKeyName` we use `objectKeyNameInvalid`.

I'll add the feedback to each item individually.

faisal154 committed:
fix: error message to camel case

faisal154 committed:
fix: revert testcase for containedBy invalid query

faisal154 committed:
update reqOp and noOp message The -> The field

faisal154 committed:
fix: refactor 1

faisal154 committed:
fix: Error Message refactor

faisal154 committed:
fix: import ErrorMessage

faisal154 committed:
fix: import ErrorMessage -- 2

faisal154 committed:
fix: ValidationAndPasswordsReset spec file fix

faisal154 committed:
fix: use ErrorMessage object in ValidationAndPasswordReset spec

faisal154 committed:
fix: test fix for EmailVerificationToken spec

faisal154 committed:
fix: failing test fixes for ParseUser spec

faisal154 committed:
fix: failing test cases AccountLockoutPolicy spec

faisal154 committed:
fix: refactor accountLockout string to ErrorMessage

faisal154 committed:
fix: failing test case for ParseServerRESTController spec

faisal154 committed:
fix: schema spec failing test cases

faisal154 committed:
fix: failing test cases for VerifyUserPassword spec

faisal154 committed:
fix: failing test cases rest spec

faisal154 committed:
fix: RestQuery spec failing test non-exitent class

faisal154 committed:
fix: RestQuery to use ErrorMessage for class creation err

faisal154 committed:
fix: ParseQyery spec

faisal154 committed:
fix: ParseQuery.FullTextSearch test case fix

faisal154 committed:
fix: ParseQuery.FullTextSearch spec failing errors

faisal154 committed:
fix: ParseQuery.FullTextSearch spec failing errors -- 2

faisal154 committed:
fix: invalidJson test fix --1

faisal154 committed:
fix: ParseQuery.FullTextSerach spec test cases for pgsql

faisal154 committed:
fix: PostgresStorageAdapter error :

faisal154 committed:
fix: added test cases for ErrorMessage

faisal154 committed:
fix: removed test cases for ErrorMessage

fn-faisal commented:
@mtrezza I have made the updates and previously, all tests but the `codecov` ones were passing. I pushed some commits and there's one postgres test that sometimes fails.
[https://github.com/parse-community/parse-server/blob/alpha/spec/ParseGraphQLServer.spec.js#L2603](https://github.com/parse-community/parse-server/blob/alpha/spec/ParseGraphQLServer.spec.js#L2603) 

There's a note on top of the test. Should I create an issue for this?

mtrezza commented:
@fn-faisal some tests are flaky, we can just ignore them in this PR.

mtrezza commented on spec/ParseQuery.FullTextSearch.spec.js:
```diff
@@ -110,7 +111,10 @@ describe('Parse.Query Full Text Search testing', () => {
-      new Parse.Error(Parse.Error.INVALID_JSON, 'bad $text: $search, should be object')
+      new Parse.Error(
+        Parse.Error.INVALID_JSON,
+        ErrorMessage.queryValueTypeInvalid('object', '$text: $search,')
```
this looks odd, why is a comma needed at the end of the 2nd argument `$text: $search,`? Maybe we should move that into the error message?

fn-faisal replied:
I have moved them to the error message.

mtrezza commented on spec/ParseQuery.spec.js:
```diff
@@ -980,7 +981,7 @@ describe('Parse.Query testing', () => {
       .catch(response => {
         equal(response.data.code, Parse.Error.INVALID_JSON);
-        equal(response.data.error, 'bad $containedBy: should be an array');
+        equal(response.data.error, ErrorMessage.queryValueTypeInvalid('an array', '$containedBy:'));
```
This also looks odd, can we move the colon in `$containedBy:` into the error message?

fn-faisal replied:
I have refactored `queryValueTypeInvalid` to include key - value args and removed the colon.

mtrezza commented on spec/ParseQuery.spec.js:
```diff
@@ -1319,7 +1320,7 @@ describe('Parse.Query testing', () => {
       .catch(error => {
         equal(error.code, Parse.Error.INVALID_JSON);
-        equal(error.message, 'bad $in value');
+        expect(['bad $in value', 'bad $in value.'].indexOf(error.message) > -1).toBeTruthy();
```
I think I mentioned this somewhere, why can't we just use `expect...contains`?

fn-faisal replied:
I have reverted this and fix the test failing issue.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,44 @@
+  notFound: type => `${type} not found.`,
+  exists: (type, key) => `${type} ${key} already exists`,
+  required: (field, op = ' to save/update the object', isField = false) =>
+    `A value for ${isField === true ? 'field ' : ''}${field} is required${op}.`,
```
This looks a bit complicated, with default values and no spacing for `op`. Can you give examples for the different parameters and which error messages they produce? I suspect this can be simplified.

fn-faisal replied:
op has a default value which has appended space to it. This message is can have an optional op appended to it. I have reverted it to check if op is provided and print a blank space before op.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,44 @@
+  geoJsonInvalidVertices: () => 'GeoJSON: Loop must have at least 3 different vertices.',
+  invalidPolygonValues: () => 'Polygon must have atleast 3 values.',
+  invalidNestedKey: () => "Nested keys should not contain the '$' or '.' characters",
+  objectFieldValueInvalid: type => `bad ${type} value.`,
```
Can you convert this into a full error message, starting capitalized? Let's avoid the word "bad" and instead use "invalid" as is used in most other messages.

fn-faisal replied:
I have refactored it Invalid ${type} value provided.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,44 @@
+  queryAllValueInvalid: value => `All $all values must be of regex type or none: ${value}.`,
+  queryValueTypeInvalid: (type, value) => `bad ${value} should be ${type}.`,
+  valueNotSupported: value =>
+    `bad ${value} not supported, please use $regex or create a separate lower case column.`,
```
Can you convert this into a full error message, starting capitalized? Let's avoid the word "bad" and instead use "invalid" as is used in most other messages.

fn-faisal replied:
I have updated this.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,44 @@
+  valueNotSupported: value =>
+    `bad ${value} not supported, please use $regex or create a separate lower case column.`,
+  databasePostgresExtensionRequired: (value, extension) =>
+    `bad ${value} not supported, install ${extension} Extension.`,
```
Same here

fn-faisal replied:
I have updated this.

faisal154 committed:
fix: refactor queryValueTypeInvalid seperator

faisal154 committed:
fix: refactor queryValueTypeInvalid key, value

faisal154 committed:
fix: failing testcases for mongo

faisal154 committed:
fix: refactor required op space

faisal154 committed:
fix: bad renamed to invalid

faisal154 committed:
fix: objectFieldValueInvalid refactor

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,45 @@
+  duplicateValue: () => 'A duplicate value for a field with unique values was provided.',
+  databaseAdapterError: () => 'Database adapter error.',
+  geoJsonInvalidVertices: () => 'GeoJSON: Loop must have at least 3 different vertices.',
+  invalidPolygonValues: () => 'Polygon must have atleast 3 values.',
```
please add space between "at"  and "least"

fn-faisal replied:
I have updated this.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,45 @@
+  objectFieldValueInvalid: type => `Invalid ${type} value provided.`,
+  queryAllValueInvalid: value => `All $all values must be of regex type or none: ${value}.`,
+  queryValueTypeInvalid: (type, key, value) =>
+    `Invalid ${key}: ${value}${value ? ',' : ''} should be ${type}.`,
```
I'm looking at his example: `queryValueTypeInvalid('string', '$text', '$language')`

- I assume that there is always a 2nd and 3rd argument, so there is no need to pass them as separate arguments.
- I assume that the argument name "value" is misleading, because the value seems to be just another key (`$language`)?

So maybe we can simplify to this:

```
queryValueTypeInvalid: (type, key) => `Value of '${key}' must be of type ${type}.`,
```

then called with:

```
queryValueTypeInvalid('string', '$text.$language');
```

fn-faisal replied:
I have updated this.

mtrezza commented on src/Errors/message.js:
```diff
@@ -0,0 +1,45 @@
+  invalidPolygonValues: () => 'Polygon must have atleast 3 values.',
+  invalidNestedKey: () => "Nested keys should not contain the '$' or '.' characters",
+  objectFieldValueInvalid: type => `Invalid ${type} value provided.`,
+  queryAllValueInvalid: value => `All $all values must be of regex type or none: ${value}.`,
```
Does this require a dedicated error message, or can we just use `queryValueTypeInvalid`?

Would this be correct?
```
queryValueTypeInvalid('regex or none', '$all');
```

(not sure what `none` is supposed to mean in this error message, but let's just use it as in the original message)

fn-faisal replied:
I have refactored it to use `queryValueTypeInvalid` instead and removed `queryAllValueInvalid`

mtrezza reviewed (commented):
Could you take a look at the previous comments, like https://github.com/parse-community/parse-server/pull/7446#discussion_r760190848. Maybe you did not see some of the comments because they were folded in the thread. These comments mostly require renaming the error keys, to make it easier to understand what they are used for. It would then also be good if you could sort the error messages alphabetically, so we have a list that is easy to read.

Thanks for the effort, this is quite a large change, so it takes some revisions.

faisal154 committed:
Merge branch 'alpha' of https://github.com/parse-community/parse-server into feat/1731

faisal154 committed:
fix: refactor error messages

faisal154 committed:
fix: refactor error message

fn-faisal commented:
> Could you take a look at the previous comments, like [#7446 (comment)](https://github.com/parse-community/parse-server/pull/7446#discussion_r760190848). Maybe you did not see some of the comments because they were folded in the thread. These comments mostly require renaming the error keys, to make it easier to understand what they are used for. It would then also be good if you could sort the error messages alphabetically, so we have a list that is easy to read.
> 
> Thanks for the effort, this is quite a large change, so it takes some revisions.
@mtrezza 
I have gone through the comments and updated the PR.

faisal154 committed:
fix: test cases postgres adapter

fn-faisal committed:
Merge branch 'alpha' into feat/1731

mtrezza commented:
@fn-faisal Apologies for the delay, I should look over this over the next days, it's quite a big (and great ;) PR.

fn-faisal commented:
@mtrezza NP 👍 . I've gone through the requested changes have updated the PR as requested. Please let me know if any other updates are required.

fn-faisal committed:
Merge branch 'alpha' into feat/1731

fn-faisal commented:
@mtrezza I have resolved some of the merge conflicts for the PR. Please let me know of any required updates for the PR once you review it. 
Thanks.

mtrezza commented:
Thanks @fn-faisal, I'll review in the coming days, apologies for the delay.

fn-faisal commented:
@mtrezza ok, sounds good. thanks.

dblythy commented:
What is pending here @mtrezza?

mtrezza commented:
It needs a rebase and a complete review of the suggested changes.

fn-faisal commented:
Hi guys, 
I made this PR about a year ago back when I was working at Back4App but as I'm no longer working with them and have moved on to better opportunities, I'm afraid I won't be able to invest time on the open issues anymore.
@mtrezza if you'd like, I'm ok with closing this PR or if someone else would like to pick this one up, that's cool too.

mtrezza commented:
Thanks for the heads up, I guess we can keep it open for someone else to continue it.
